### DISCUSSION TITLE
Using an external Node.js module, and including said module in my Visual Studio project so that it's also published

### DISCUSSION THREAD

**Discussion by njherrera** [agracio/edge-js#220] [Q&A] [answered]
Hello, this isn't exactly a question specific to just edge-js, but this seems like a good place to ask it!

I'm currently working on a C#/JS project that uses the [slippi-js](https://github.com/project-slippi/slippi-js) JavaScript library to find instances of specific situations within replay files for Super Smash Bros. Melee games. Currently I use named pipes so that my C# parser can send a request for data from specific replay files to my JS code, then receive that data in JSON form and parse it. The slippi-js library is the best option available for getting data from the replay files, so it's essentially a requirement that I use it in my project.

The named pipe system does the job, but I've been looking into using this library to write an interpreter as an alternative approach. My question is, what's the best way to include an external JS library (like slippi-js) in my project such that it can still be processed correctly by edge-js? 

The README says this, which I understand and makes sense to me.

> NPM modules must be installed in the directory where your build system places the Edge.js NuGet package (most likely the same location as the rest of your application binaries), or any ancestor directory.

However, when it comes to publishing a stand-alone version of my project, I'm not sure how I could satisfy that requirement while also including the slippi-js library. I'm also pretty unsure as to how including JS libraries in a Visual Studio C# project generally works, it's been a little hard to find good info on it.

The best lead I've found/idea I've had so far is modifying the repositoryPath in the project's nuget.config such that it puts the edge-js NuGet package into a specific packages/library folder within the project folder, then also installing the JS library into that folder. However, I'm not sure if that would accomplish what I want, or even be a good idea in the first place.

Any help would be appreciated, this has been a real head-scratcher!

**Comment by agracio** [comment, marked as answer]
There are no additional steps needed but it depends on your solution structure. Using an example from readme it would look like this:

`package.json`
```js
{
  "name": "edge-js-websockets",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "MIT",
  "description": "",
  "dependencies": {
    "ws": "^8.18.0"
  }
}
```

`Websockets.sln`

```
Microsoft Visual Studio Solution File, Format Version 12.00
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Websockets", "Websockets\Websockets.csproj", "{ECB9E1B6-4216-4506-9B8A-84326A03AE9A}"
EndProject
Global
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{ECB9E1B6-4216-4506-9B8A-84326A03AE9A}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{ECB9E1B6-4216-4506-9B8A-84326A03AE9A}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{ECB9E1B6-4216-4506-9B8A-84326A03AE9A}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{ECB9E1B6-4216-4506-9B8A-84326A03AE9A}.Release|Any CPU.Build.0 = Release|Any CPU
	EndGlobalSection
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Any CPU = Debug|Any CPU
		Release|Any CPU = Release|Any CPU
	EndGlobalSection
EndGlobal
```

`Websockets\Websockets.csproj`

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net462</TargetFramework>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <AssemblyTitle>Websockets</AssemblyTitle>
    <Product>Websockets</Product>
    <Copyright>Copyright ©  2024</Copyright>
    <OutputPath>bin\$(Configuration)\</OutputPath>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <DebugType>full</DebugType>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <DebugType>pdbonly</DebugType>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="EdgeJs" Version="20.12.3" />
  </ItemGroup>
  <ItemGroup>
    <Content Include="edge\double_edge.js" />
    <Content Include="edge\edge.js" />
    <Content Include="edge\x64\edge_nativeclr.node" />
    <Content Include="edge\x64\libnode.dll" />
    <Content Include="edge\x86\edge_nativeclr.node" />
    <Content Include="edge\x86\libnode.dll" />
  </ItemGroup>
</Project>

```

`Websockets\Program.cs`

```cs
using System;
using System.Threading;
using System.Threading.Tasks;
using EdgeJs;

namespace Websockets
{
    internal class Program
    {
        public static async void Start()
        {
            var createWebSocketServer = Edge.Func(@"
            var WebSocketServer = require('ws').Server;

            return function (port, cb) {
                var wss = new WebSocketServer({ port: port });
                wss.on('connection', function (ws) {
                    ws.on('message', function (message) {
                        ws.send(message.toString().toUpperCase());
                    });
                    ws.send('Hello!');
                });
                cb();
            };
        ");

            await createWebSocketServer(8080);
        }

        static void Main(string[] args)
        {
            Task.Run((Action)Start);
            new ManualResetEvent(false).WaitOne();
        }
    }
}
```

When deploying solution you would need to copy `node_modules` folder to its correct location depending on your solution structure, you would not need `package.json` in your deployed application.


Based on the above example deployed application would have `node_modules` and everything from `Websockets\bin\Debug`.

`C:\websockets-deploy\`

```
edge (folder)
net462 (folder)
node_modules (folder)
EdgeJs.dll
Websockets.exe
Websockets.pdb
```

**Reply by njherrera** [reply]
Thank you very much for the answer! It helps a lot to have it laid out like this. 

Just so I can better understand the details, what's the purpose of the package.json file here? Is it purely for putting together the node_modules folder so that edge-js can find any imported modules? Also if so, is that the way JS InterOp libraries like edge-js generally work - looking for any required libraries in a node_modules folder within the project/solution structure?

Thanks again for the help, I very much appreciate it!

**Reply by agracio** [reply]
`package.json` is needed in your project to install all your project Node.js dependencies, in this case it is `"ws": "^8.18.0"` while in your case it would be `"@slippi/slippi-js": "^6.7.0"`. Running `npm i` in root directory will then add all dependant node modules. 

`node_modules` folder contains all the libraries that your Node.js code is trying to access using `require('')`, any such call would try to find relevant module in `node_modules` folder. This is not specific to interpreter libraries, it is a standard Node.js module resolution behaviour.

In the example code `var WebSocketServer = require('ws').Server;` would trigger Node.js to look for entry point for `ws`  in `node_modules/ws`.
In your code you would use `const { SlippiGame } = require("@slippi/slippi-js");` triggering similar action from Node.js.

**Reply by njherrera** [reply]
That makes total sense, I guess I've just never looked into how Node.js resolves `require` statements but it feels a lot more straightforward now. Thanks again, the in-depth explanation was very helpful!
